### DISCUSSION TITLE
Changing hardware serial parameters to 7E1

### DISCUSSION THREAD

**Discussion by Andreas2430** [arendst/Tasmota#15398] [Q&A] [answered]
Hi,

the tasmota documentation for smart metering [explains](https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3d-q3da1024-obis) how TasmotaSerial.cpp has to be modified to enable "unusual" hardware serial parameters like 7E1.

The modifications proposed in the above link apply to an older version (3.2.0) of TasmotaSerial.cpp.

How does the current version 3.5.0 of TasmotaSerial.cpp have to be modified to achieve the same functionality?

Thanks!
Kind regards
Andreas

**Comment by barbudor** [comment, marked as answer]
Hi Andreas
You shouldn't change Tasmota code for that
If you are using Smart Meter feature with the Scripting Language read the end of this section:

https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-definition

**Reply by Andreas2430** [reply]
Hi barbudor, 
thanks for the hint! I will try that when my hardware arrives.
If it works correctly, then the [text in the Tasmota SML documentation regarding the EasyMeter Q3D](https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3d-q3da1024-obis) could be updated to reflect that users do not need to change the TasmotaSerial.cpp code anymore to enable 7E1.

**Reply by stefani** [reply]
Hi Andreas, 

did you find a solution? 

The documentation says I have to change source code, but the code for version 3.2.0 does not fit into code with version 3.5.0. There is no documented option to change parameters from 8N1 to 7E1. You can modify parity and stop bits, but there is no description how to modify data bits (from 8 to 7).

How did you managed this?

Kind regards
Sven

**Reply by barbudor** [reply]
@stefani you are correct, parity and stop bits only
Modifying the source code of TasmotaSerial.cpp is a real bad idea and the code has evolved, so the documentation is however outdated

@gemu2015 Could you please advise if a better solution could be used ?

**Reply by Andreas2430** [reply]
Hi @stefani, the solution was actually really simple, but I only stumbled upon it by accident.

I am using a WiFi Infrared reader based on an ESP01s, and the manufacturer of this device has uploaded a Youtube video explaining how to test, configure and operate the device. [At one moment in this video](https://youtu.be/VuXpzKetOhc?t=706), he says that when you configure the meter type in the [meter definition](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-definition) of Tasmota's smart meter script to "OBIS", Tasmota will automatically set the serial connection to 7E1. 

This worked fine for me. There was no source code modification necessary.

My functioning Tasmota script code (for an EasyMeter Q1DB1004) is:

```
>D
>B
=>sensor53 r
>M 1
+1,3,o,0,9600,SML
1,1-0:1.8.0*255(@1,EC_CounterVal,kWh,EC_CounterVal,4
1,1-0:61.7.255*255(@1,EC_PowerVal,W,EC_PowerVal,0
#
```

and the "o" in line 5 is where this magic happens. I don't know if that feature is explained somewhere in Tasmota's documentation - I couldn't find it there. 

Hope this helps in your usecase, too!

**Reply by stefani** [reply]
Hello!

@barbudor yes, I agree, modifying the source code is a bad idea (if there is another way).

@Andreas2430 thanks for your detailed reply. I did exactly this multiple times ago. With tasmota 9.5. And now suddenly it works. I don't now why. Maybe the receiver and the sensor weren't properly aligned. My first attempts resulted in something like 

> ^3f7<8?x~n|n<C{C{C{C{{cO^8~C[87|on<C{o{{O?8{cO^8~C[87|on<C{O?|?8{cCCCc|

so i assumed it was the wrong connection parameters.

**Comment by barbudor** [comment]
The doc is open source too and can be edits can be proposed by clicking on the pencil icon on the top right
😂

**Reply by Andreas2430** [reply]
Found the edit button now, thanks :)
